dali/pipeline/operators/reader/nvdecoder/nvdecoder.h

```diff
@@ -0,0 +1,209 @@
+    CUStream& operator=(const CUStream&) = delete;
+    CUStream(CUStream&&);
+    CUStream& operator=(CUStream&&);
+    operator cudaStream_t();
+
+  private:
+    bool created_;
+    cudaStream_t stream_;
+};
+
+
+enum ScaleMethod {
+    /**
+     * The value for the nearest neighbor is used, no interpolation
+     */
+    ScaleMethod_Nearest,
+
+    /**
+     * Simple bilinear interpolation of four nearest neighbors
+     */
+    ScaleMethod_Linear
+};
+
+class NvDecoder
+{
+  public:
+    NvDecoder(int device_id,
+              const CodecParameters* codecpar,
+              AVRational time_base);
+
+    NvDecoder(const NvDecoder&) = default;
+    NvDecoder(NvDecoder&&) = default;
+    NvDecoder& operator=(const NvDecoder&) = default;
+    NvDecoder& operator=(NvDecoder&&) = default;
+    ~NvDecoder();
+
+    bool initialized() const;
+
+    static int CUDAAPI handle_sequence(void* user_data, CUVIDEOFORMAT* format);
+    static int CUDAAPI handle_decode(void* user_data, CUVIDPICPARAMS* pic_params);
+    static int CUDAAPI handle_display(void* user_data, CUVIDPARSERDISPINFO* disp_info);
+
+    int decode_packet(AVPacket* pkt);
+
+    void push_req(FrameReq req);
+
+    void receive_frames(SequenceWrapper& batch);
+
+    void finish();
+
+  protected:
+    int decode_av_packet(AVPacket* pkt);
+
+    void record_sequence_event_(SequenceWrapper& sequence);
+
+    const int device_id_;
+    CUStream stream_;
+    const CodecParameters* codecpar_;
+
+  private:
+
+    class MappedFrame {
+      public:
+        MappedFrame();
+        MappedFrame(CUVIDPARSERDISPINFO* disp_info, CUvideodecoder decoder,
+                    CUstream stream);
+        ~MappedFrame();
+        MappedFrame(const MappedFrame&) = delete;
+        MappedFrame& operator=(const MappedFrame&) = delete;
+        MappedFrame(MappedFrame&& other);
+        MappedFrame& operator=(MappedFrame&&) = delete;
+
+        uint8_t* get_ptr() const;
+        unsigned int get_pitch() const;
+
+        CUVIDPARSERDISPINFO* disp_info;
+      private:
+        bool valid_;
+        CUvideodecoder decoder_;
+        CUdeviceptr ptr_;
+        unsigned int pitch_;
+        CUVIDPROCPARAMS params_;
+    };
+
+    class TextureObject {
+      public:
+        TextureObject();
+        TextureObject(const cudaResourceDesc* pResDesc,
+                      const cudaTextureDesc* pTexDesc,
+                      const cudaResourceViewDesc* pResViewDesc);
+        ~TextureObject();
+        TextureObject(TextureObject&& other);
+        TextureObject& operator=(TextureObject&& other);
+        TextureObject(const TextureObject&) = delete;
+        TextureObject& operator=(const TextureObject&) = delete;
+        operator cudaTextureObject_t() const;
+      private:
+        bool valid_;
+        cudaTextureObject_t object_;
+    };
+
+    struct TextureObjects {
+        TextureObject luma;
+        TextureObject chroma;
+    };
+
+    CUdevice device_;
+    CUContext context_;
```

Comment:
Do we use it anywhere?

Reply:
Yes here for instance:
https://github.com/NVIDIA/DALI/blob/740bef6d89d55326e5f1c5ee0b8649026188a6de/dali/pipeline/operators/reader/nvdecoder/nvdecoder.cc#L155